File: Linked Lists/LinkedList.py
from asyncio.windows_events import NULL


class Node:
    
    def __init__(self, data = 0, nextNode = NULL):
        self.__data = data
        self.__nextNode = nextNode
    
    def getNextNode(self):
        return self.__nextNode
    def setNextNode(self, newNextNode):
        self.__nextNode = newNextNode
    def getData(self):
        return self.__data
    def setData(self, newData):
        self.__data = newData

    def __str__(self):
        output = "Data: "
        output += self.__data + "\nNext Node: "
        if (self.__nextNode != NULL):
            output += self.__nextNode.getData()
        else:
            output += "NULL"
        return output

class LinkedList:

    def __init__(self, firstNode = NULL):
        self.__head = firstNode
        self.__tail = firstNode
        if (firstNode == NULL):
            self.__size = 0
        else:
            self.__size = 1

    def setHead(self, newHeadNode):
        currentHead = self.__head
        newHeadNode.setNextNode(currentHead)
        self.__head = newHeadNode
        self.__size += 1
    def getHead(self):
        return self.__head
    def append(self, newTailNode):
        self.__tail.setNextNode(newTailNode)
        self.__tail = newTailNode
        self.__size += 1
    def getTail(self):
        return self.__tail
    def getSize(self):
        return self.__size

    def getNode(self, index):
        if (index < self.__size):
            output = self.__head
            for i in range(index):
                output = output.getNextNode()
            return output
        else:
            raise IndexError("INDEX OUT OF BOUNDS")

    def insert(self, index, newNode):
        if (index < self.__size):
            if (index == 0):
                self.setHead(newNode)
            elif (index == self.__size):
                self.append(newNode)
            else:
                nextNode = self.getNode(index)
                prevNode = self.getNode(index - 1)
                prevNode.setNextNode(newNode)
                newNode.setNextNode(nextNode)
                self.__size += 1
        else:
            raise IndexError("INDEX OUT OF BOUNDS")
    
    def pop(self, index):
        output = NULL
        if ((index < (self.__size - 1)) and (index > 0)):
            prev = self.getNode(index - 1)
            output = self.getNode(index)
            prev.setNextNode(self.getNode(index + 1))
            self.__size -= 1
        elif (index == 0):
            output = self.__head
            self.__head = self.__head.getNextNode()
            self.__size -= 1
        elif (index == (self.__size - 1)):
            output = self.getNode(index)
            prev = self.getNode(index - 1)
            prev.setNextNode(NULL)
            self.__size -= 1
        return output

    def splice(self, startingIndex, length):
        print("FINISH THIS")


    def reverse(self):
        if (self.__size > 1):
            start = self.__head
            prev = self.__head
            current = self.__head.getNextNode()
            for i in range(self.__size - 1):
                next = current.getNextNode()
                current.setNextNode(prev)
                prev = current
                current = next
            self.__head = prev
            start.setNextNode(NULL)


    def __str__(self):
        output = ""
        for i in range(self.__size):
            if (i != (self.__size - 1)):
                output += str(self.getNode(i).getData()) + " ->\n"
            else:
                output += str(self.getNode(i).getData()) + " (END)"
        return output

list = LinkedList(Node("Twenty"))
list.append(Node("TwentyOne"))
list.append(Node("TwentyTwo"))
list.setHead(Node("Nineteen"))
list.insert(3, Node("TwentyOnePointFive"))
list.insert(1, Node("HELLO"))
list.pop(1)

print(list)
list.reverse()
print(list)

print()
print(list.getNode(2))

File: Linked-Lists/Java/LinkedList.java
//Gregory Ecklund
//September 2022

public class LinkedList {
    /**
        LinkedList Class - Data Structure

        Fields:
            private Node head (default = startingNode/null) - The first node in the LinkedList
            private Node tail (default = startingNode/null) - The last node in the LinkedList
            private int size (default = 0/1) - The size of the LinkedList
    */

    private Node head;
    private Node tail;
    private int size;

    public LinkedList() {
        /**
            Constructor() - Base constructor, no parameters
        */
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    public LinkedList(Node startingNode) {
        /**
            Constructor(Node startingNode) - Constructor containing a single starting Node

            Parameters:
                Node startingNode - The desired initial first Node in the LinkedList
        */

        this.head = startingNode;
        this.tail = startingNode;
        this.size = 1;
    }

    public void setHead(Node newHead) {
        /**
            setHead(Node newHead) - Adds newHead to the start of the LinkedList

            Parameters:
                Node newHead - The Node you wish to be the new Head
        */

        Node oldHead = this.head;
        this.head = newHead;
        this.head.setNextNode(oldHead);
        this.size++;
    }

    public Node getHead() {
        /**
            getHead() - Gets the head Node of the LinkedList

            Returns:
                Node - Returns the head field of the LinkedList
        */

        return this.head;
    }

    public void append(Node newTail) {
        /**
            append(Node newTail) - Adds newTail to the end of the LinkedList

            Parameters:
                Node newTail - The new Node you wish to add to the LinkedList
        */

        if (this.size != 0) {
            this.tail.setNextNode(newTail);
            this.tail = newTail;
        }
        else {
            this.head = newTail;
            this.tail = newTail;
        }
        this.size++;
    }

    public Node getTail() {
        /**
            getTail() - Gets the tail Node of the LinkedList

            Returns:
                Node - Returns the tail field of the LinkedList
        */

        return this.tail;
    }

    public int getSize() {
        /**
            getSize() - Gets the size of the LinkedList

            Returns:
                int - Returns the size field of the LinkedList
        */

        return this.size;
    }

    public Node getNode(int index) {
        /**
            getNode(int index) - Gets the Node at the requested index of the LinkedList

            Parameters:
                int index - The index of the Node you would like to get from LinkedList
            
            Returns:
                Node - Returns the Node at the requested index of LinkedList

            IndexOutOfBoundsException:
                index argument is out of bounds for length of LinkedList
        */

        if (index < this.size && index > -1) {
            Node output = this.head;
            for (int i = 0; i < index; i++) {
                output = output.getNextNode();
            }
            return output;
        }
        else {
            throw new IndexOutOfBoundsException("index argument is out of bounds for length of LinkedList");
        }
    }

    public void insert(int index, Node node) {
        /**
            insert(int index, Node node) - Inserts node at the requested index of the LinkedList

            Parameters:
                int index - The index of the LinkedList you wish to insert node into
                Node node - The new Node you wish to insert into LinkedList at requested index
            
            IndexOutOfBoundsException:
                index argument is out of bounds for length of LinkedList
        */
    
        if (index < this.size && index > -1) {
            if (index == 0) {
                this.setHead(node);
            }
            else if (index == this.size) {
                this.append(node);
            }
            else {
                Node oldNode = this.getNode(index);
                Node prevNode = this.getNode(index - 1);
                prevNode.setNextNode(node);
                node.setNextNode(oldNode);
                this.size++;
            }
        }
        else {
            throw new IndexOutOfBoundsException("index argument is out of bounds for length of LinkedList");
        }
    }

    public Node pop() {
        /**
            pop() - Removes and returns the tail from LinkedList

            Returns:
                Node - Returns the tail
        */

        return pop(this.size - 1);
    }

    public Node pop(int index) {
        /**
            pop(int index) - Removes and returns the Node from LinkedList at requested index

            Parameters:
                int index - The index of the Node you wish to remove and return from LinkedList

            Returns:
                Node - Returns the Node at the requested index of LinkedList
            
            IndexOutOfBoundsException:
                index argument is out of bounds for length of LinkedList
        */

        Node output = null;
        if ((index < (this.size - 1)) && (index > 0) && (this.size != 0)) {
            Node prev = this.getNode(index - 1);
            output = this.getNode(index);
            prev.setNextNode(this.getNode(index + 1));
            this.size--;
        }
        else if (index == 0 && this.size != 0) {
            output = this.head;
            this.head = this.head.getNextNode();
            this.size--;
        }
        else {
            throw new IndexOutOfBoundsException("index argument is out of bounds for length of LinkedList");
        }
        return output;
    }

    public void reverse() {
        /**
            reverse() - Reverses the order of the LinkedList
        */

        if (this.size > 1) {
            Node start = this.head;
            Node prev = this.head;
            Node current = this.head.getNextNode();
            Node next;
            for (int i = 0; i < (this.size - 1); i++) {
                next = current.getNextNode();
                current.setNextNode(prev);
                prev = current;
                current = next;
            }
            this.head = prev;
            start.setNextNode(null);
        }
    }

    public String toString() {
        /**
            To String Method

            Returns:
                Returns a string representation of the LinkedList
        */

        String output = "";
        Node currentNode = this.head;
        for (int i = 0; i < this.size; i++) {
            output += currentNode.getData();
            if (i != (this.size - 1)) {
                output += " ->\n";
            }
            else {
                output += " (END)";
            }
            currentNode = currentNode.getNextNode();
        }
        return output;
    }
}